packages/x-data-grid-pro/src/hooks/features/serverSideData/useGridServerSideCache.ts

```diff
@@ -0,0 +1,104 @@
+import * as React from 'react';
+import { useGridApiMethod } from '@mui/x-data-grid';
+import { GridPrivateApiPro } from '../../../models/gridApiPro';
+import { DataGridProProcessedProps } from '../../../models/dataGridProProps';
+import { GridGetRowsParams, GridGetRowsResponse, GridServerSideCache } from '../../../models';
+import { GridServerSideCacheApi } from './serverSideInterfaces';
+
+class SimpleServerSideCache {
+  private cache: Record<string, GridGetRowsResponse>;
+
+  constructor() {
+    this.cache = {};
+  }
+
+  static getKey(params: GridGetRowsParams) {
+    return JSON.stringify([
+      params.paginationModel,
+      params.filterModel,
+      params.sortModel,
+      params.groupKeys,
+    ]);
+  }
+
+  set(key: string, value: GridGetRowsResponse) {
+    this.cache[key] = value;
+  }
+
+  get(key: string) {
+    return this.cache[key];
+  }
+
+  clear() {
+    this.cache = {};
+  }
+}
+
+const getDefaultCache = (cacheInstance: SimpleServerSideCache): GridServerSideCache => ({
+  getKey: SimpleServerSideCache.getKey,
+  set: (key: string, value: GridGetRowsResponse) =>
+    cacheInstance.set(key as string, value as GridGetRowsResponse),
+  get: (key: string) => cacheInstance.get(key as string),
+  clear: () => cacheInstance.clear(),
+});
+
+export const useGridServerSideCache = (
+  privateApiRef: React.MutableRefObject<GridPrivateApiPro>,
+  props: Pick<
+    DataGridProProcessedProps,
+    'unstable_dataSource' | 'disableServerSideCache' | 'unstable_serverSideCache'
+  >,
+): void => {
+  const defaultCache = React.useRef(getDefaultCache(new SimpleServerSideCache()));
+  const cache = React.useRef<GridServerSideCache>(
+    props.unstable_serverSideCache || defaultCache.current,
+  );
+
+  const getCacheData = React.useCallback(
+    (params: GridGetRowsParams) => {
+      if (props.disableServerSideCache) {
+        return undefined;
+      }
+      const key = cache.current.getKey(params);
+      return cache.current.get(key);
+    },
+    [props.disableServerSideCache],
+  );
+
+  const setCacheData = React.useCallback(
+    (params: GridGetRowsParams, data: GridGetRowsResponse) => {
+      if (props.disableServerSideCache) {
+        return;
+      }
+      const key = cache.current.getKey(params);
+      cache.current.set(key, data);
+    },
+    [props.disableServerSideCache],
+  );
+
+  const clearCache = React.useCallback(() => {
+    if (props.disableServerSideCache) {
+      return;
+    }
+    cache.current.clear();
+  }, [props.disableServerSideCache]);
+
+  const serverSideCacheApi: GridServerSideCacheApi = {
+    getCacheData,
+    setCacheData,
+    clearCache,
+  };
+
+  useGridApiMethod(privateApiRef, serverSideCacheApi, 'public');
+
+  const isFirstRender = React.useRef(true);
+  React.useEffect(() => {
+    if (isFirstRender.current) {
+      isFirstRender.current = false;
+      return;
+    }
```

Comment:
Not sure but we might have a hook that does this already, in some utils package/folder.

Reply:
Do you mean [`useFirstRender`](https://github.com/mui/mui-x/blob/c57f7644d38d500031b46a4000a489aa96f22621/packages/x-data-grid/src/hooks/utils/useFirstRender.ts#L3)?

If yes, then I think we still need to [use](https://github.com/mui/mui-x/blob/c57f7644d38d500031b46a4000a489aa96f22621/packages/x-data-grid/src/hooks/core/strategyProcessing/useGridRegisterStrategyProcessor.ts#L19-L31) a local ref if we want to do something on subsequent renders.